Document how to split an image from an existing repo (#858)
2i2c-org/infrastructure: docs/howto/operate/ci-cd/split-repo2docker.md

# Split up an image for use with r2d-action

Sometimes we have user images defined in a repo, and we want to extract
it to a standalone repo so it can be used with the [repo2docker action](https://github.com/jupyterhub/repo2docker-action).
But we want to retain full history of the image still, so we can look
back and see why things are the way they are, as well as credit the people
who contributed over time. This page documents the git-fu required to
make this happen.

1. Clone the base repository you are extracting the image from.
   We will be performing destructive operations on this clone,
   so it has to be a fresh clone.

    ```bash

>     git clone https://github.com/utoronto-2i2c/jupyterhub-deploy utoronto-image --origin source

Reviewer: ```suggestion
    git clone <base-repo-url> <hub-image> --origin source
```

    ```

>     We name the directory `utoronto-image` as that would be the name of the

Reviewer: ```suggestion
    We name the directory `<hub-image>` as that would be the name of the
```

    repo with just the user image. We ask git to name the remote it creates be
    called `source`, as we will create a new GitHub repo later that will be `origin`.

> 2. Make sure [git-filter-repo](https://github.com/newren/git-filter-repo)

Reviewer 2: IIRC, I have used `git subtree split` in the past and I believe it already comes with git...
Do you prefer the `git-filter-repo` tool because of some additional features it has?
Just curious 😉 !

Author: I've used `git filter-branch` before, and if you try it now, it prints out this official warning:

```
➜ git filter-branch
WARNING: git-filter-branch has a glut of gotchas generating mangled history
         rewrites.  Hit Ctrl-C before proceeding to abort, then use an
         alternative filtering tool such as 'git filter-repo'
         (https://github.com/newren/git-filter-repo/) instead.  See the
         filter-branch manual page for more details; to squelch this warning,
         set FILTER_BRANCH_SQUELCH_WARNING=1.
```

so I decided to take their advice and use filter-repo :D I didn't know subtree split could be used to do this! I've generally never used subtree...

>    is installed. It should be available in [brew or other package managgers]
>    (https://github.com/newren/git-filter-repo/blob/main/INSTALL.md)

Reviewer 3: ```suggestion
   is installed. It should be available in [brew or other package managers](https://github.com/newren/git-filter-repo/blob/main/INSTALL.md)
```

3. Run `git-filter-repo` to remove everything from the repo except the
   image directory.

   ```bash

>     git filter-repo --subdirectory-filter deployments/utoronto/image --force

Reviewer: ```suggestion
    git filter-repo --subdirectory-filter <path-to-image-directory> --force
```

    ```

>     The repo root directory now contains the contents of `deployments/utoronto/image`,

Reviewer: ```suggestion
    The repo root directory now contains the contents of `<path-to-image-directory>`,
```

    as well as full git history for any commits that touched it! This way,

>     we do not lose history or attribution.

Reviewer: It's amazing that we can still have the history of changes!

> 4. Create a new repository in GitHub based off [this template](https://github.com/2i2c-org/hub-user-image-template).

Reviewer 3: ```suggestion
4. Create a user image repository in GitHub based off [this template](https://github.com/2i2c-org/hub-user-image-template).
```

   The template makes it much easier to setup repo2docker-action on it.

> 5. Add the new repo as a remote you can push commits to / pull commits from.

Reviewer 3: ```suggestion
5. Add the new user image repo as a remote you can push commits to / pull commits from.
```

just to make it more explicit which repo we're talking about

   This will be the primary location of this repo now, so let's call it `origin`.

   ```bash

>    git remote add origin git@github.com:2i2c-org/utoronto-image.git

Reviewer: ```suggestion
   git remote add origin git@github.com:2i2c-org/<hub-image>.git
```

Reviewer 4: Does it have to be the 2i2c-org? I think this command could be:

```
git remote add origin git@github.com:<your-github-username-or-org>/<hub-image>.git
```

Author: I've done what @sgibson91 suggested here.

   ```

6. Fetch the new repo and check out the `main` branch, as that is the final
   destination for our image contents.

   ```bash
    git fetch origin
    git checkout main
    ```

7. Remove the `environment.yml` file from the repo - it is present as an example
   only, and we have our own we are bringing.

   ```bash
   git rm environment.yml
   git commit -m 'Remove unused environment.yml file'
   ```

8. Merge the branch we prepared earlier with just our image contents into this
   `main` branch, while telling git to not freak out about them being from different
   repos initially.

   ```bash

>    git merge staging --allow-unrelated-histories  -m 'Bringing in image directory from deployment repo'

Reviewer 2: This is assuming the incoming branch is `staging`, right?
Maybe we need to use add a note about a possible different name? Or prepare the above "filter"ed repo to have some canonical name we can use here?

Reviewer 2: Ja, the note is just below... I should keep reading before writing... 🤦
